From 0ea293993b627e2953a1f1ce256b96069ae83e09 Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 24 Jul 2014 18:16:13 -0400
Subject: [fs] nfs: Don't reset pg_moreio in __nfs_pageio_add_request

Message-id: <1406225773-16395-11-git-send-email-steved@redhat.com>
Patchwork-id: 86638
O-Subject: [RHEL7.1 PATCH 10/10] NFS: Don't reset pg_moreio in __nfs_pageio_add_request
Bugzilla: 1120728
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

Once we've started sending unstable NFS writes, we do not want to
clear pg_moreio, or we may end up sending the very last request as
a stable write if the commit lists are still empty.

Do, however, reset pg_moreio in the case where we end up having to
recoalesce the write if an attempt to use pNFS failed.

Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upstream Id: f563b89b182594f827b4100bd34f916339785a77
Bugzilla: 1120728

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/pagelist.c b/fs/nfs/pagelist.c
index 11972a8..68b26c0 100644
--- a/fs/nfs/pagelist.c
+++ b/fs/nfs/pagelist.c
@@ -930,7 +930,6 @@ static int __nfs_pageio_add_request(struct nfs_pageio_descriptor *desc,
    nfs_pageio_doio(desc);
    if (desc->pg_error < 0)
     return 0;
-   desc->pg_moreio = 0;
    if (desc->pg_recoalesce)
     return 0;
    /* retry add_request for this subreq */
@@ -977,6 +976,7 @@ static int nfs_do_recoalesce(struct nfs_pageio_descriptor *desc)
   desc->pg_count = 0;
   desc->pg_base = 0;
   desc->pg_recoalesce = 0;
+  desc->pg_moreio = 0;
 
   while (!list_empty(&head)) {
    struct nfs_page *req;
-- 
1.7.1